### ISSUE TITLE
Investigate leaf decay causing heavy lag

### ISSUE THREAD

**Issue by williewillus** [TeamTwilight/twilightforest#17]
Due to tree generators not setting the CHECK_DECAY property to false on leaf blockstates

**Comment by Benimatic** [comment]
I fixed several bugs that were causing leaf blocks to overwrite their supporting branches. I have not observed any leaf decay since, although I haven't tested it in code or anything.

**Comment by Benimatic** [comment]
Also I did set the CHECK_DECAY to false on all the tree code I found. The flag did not seem to be working, but I didn't figure out why that was.

**Comment by Benimatic** [comment]
As an update on this issue, the CHECK_DECAY gets set to false whenever an adjacent leaf block is replaced, even during generation. Branches replacing leaves happens frequently during tree generation, and can also happen when two trees generate close together. Because of this, a large proportion of leaves have the CHECK_DECAY flag un-set when a chunk is generated, and thus it takes about a minute after a chunk is generated for the leaves to settle down.

**Comment by williewillus** [comment]
does this happen in vanilla / why doesn't it? Does vanilla generate branches then leaves around them? Maybe we can do the same.

**Comment by Benimatic** [comment]
For vanilla, it doesn't happen with individual trees, but it can happen when two trees generate close together.

In vanilla, the tree generation algorithm determines the tree geometry in memory first and then "paints" the entire tree, IIRC top-to-bottom, placing each block only once. Most of the Twilight Forest trees use a geometrical approach where we place first the trunk, some leaves, then branches with leaves on the end. Sometimes branches will go through where we've previously placed leaves, causing sections of leaves to un-flag.

We could improve our generation algorithms. The simplest way would be to just use a two-step process where we first generate all the branches, keeping the locations of future leaves in memory, and then go through and add all the leaves later. This would use more memory, and possibly more CPU time during generation, it seems like, but might pay off overall if it reduced tick load on the server later.

I'd like to have some sort of benchmark or test set up where we could generate a set number of chunks with a set seed to compare algorithms to make sure we're really improving things.

It's worth noting that changing generation methods won't fix the situation with multiple trees overlapping. Since the dimension generates higher tree density than vanilla, some leaf replacement here is almost inevitable.

Another approach would be to alter the leaf code that un-flags nearby leaves, although this might be somewhat complex, as IIRC replacement calls the onBlockRemoved method without any indication of what's being replaced or why.

**Comment by Benimatic** [comment]
![2017-07-06_17 12 29](https://user-images.githubusercontent.com/1783934/27933164-53b5ce64-626e-11e7-83cb-36794ac7d284.png)
Further investigations! Here's a screenshot with some debug code that marks all leaves with the check_decay flag as magenta.

My conclusions:

- The canopy trees are 95% good
- Most of the flagged leaves on the canopy trees are from trees intersecting, or particularly canopy mushrooms intersecting with the trees
- The hollow trees are less good. Sometimes it looks like whole chunks are flagged
- Also the leaf dungeons are flagging giant swaths of leaves

I'll see what I can do about the leaf dungeons. It may also be fairly simple to generate leaves after branches on the hollow trees, as leaves are a different structure component than the branches.

![2017-07-06_17 16 10](https://user-images.githubusercontent.com/1783934/27933462-8d8d29d8-626f-11e7-8fe4-17b55aab1445.png)

From below, you can see that every canopy tree has some flagged leaves, due to branches cutting through existing leaves. We could work to fix that.

**Comment by Benimatic** [comment]
![2017-07-06_17 24 24](https://user-images.githubusercontent.com/1783934/27933559-f85d018e-626f-11e7-8265-dcdde0f2efbb.png)

Also these oak trees in the oak savannah biome have some flagging issues.

**Comment by Benimatic** [comment]
Okay, with the two commits above, leaves should be relatively quiet in the forest.

**Comment by Drullkus** [comment]
Seems this is done?
